syntax = "proto3";

package wma.streaming;

option py_generic_services = true;

// Audio batch containing complete 4-second OGG/Opus file
message AudioBatch {
    bytes ogg_data = 1;                 // Complete 4-second OGG/Opus file
    uint64 start_ts_ms = 2;             // Batch start timestamp (epoch ms)
    uint32 duration_ms = 3;             // Batch duration (4000ms)
    string meeting_id = 4;              // Meeting identifier
    string session_id = 5;              // Session identifier
    string chunk_id = 6;                // Unique chunk identifier
    uint32 sample_rate = 7;             // Sample rate (48000)
    uint32 channels = 8;                // Number of channels (1)
    uint32 bit_rate = 10;               // Bit rate (64000)
    string codec = 11;                  // Codec name ("opus")
    string container = 12;              // Container format ("ogg")
    uint32 frame_count = 13;            // Number of 20ms frames in the OGG file (~200)
}

// Single participant crop data from video frame
message ParticipantCrop {
    string participant_id = 1;          // Participant identifier
    bytes image_data = 2;               // JPEG encoded crop image
    uint32 bbox_x = 3;                  // Bounding box X coordinate
    uint32 bbox_y = 4;                  // Bounding box Y coordinate  
    uint32 bbox_width = 5;              // Bounding box width
    uint32 bbox_height = 6;             // Bounding box height
    float confidence = 7;               // Detection confidence score
    uint64 timestamp_ms = 8;            // Frame timestamp (epoch ms)
    uint64 sequence_number = 9;         // Frame sequence number
}

// Collection of participant frames for batched sending (2s worth)
message ParticipantFrame {
    string participant_id = 1;          // Participant identifier
    repeated ParticipantCrop crops = 2;  // All crops for this participant (~10 frames)
    uint64 start_ts_ms = 3;             // Batch start timestamp (epoch ms)
    uint64 end_ts_ms = 4;               // Batch end timestamp (epoch ms)
    string meeting_id = 5;              // Meeting identifier  
    string session_id = 6;              // Session identifier
    string chunk_id = 7;                // Unique chunk identifier
    uint32 frame_count = 8;             // Number of frames in batch
    uint32 original_frame_width = 9;    // Original capture frame width
    uint32 original_frame_height = 10;  // Original capture frame height
}

// Banner level enumeration
enum BannerLevel {
    BANNER_NONE = 0;        // No banner / hide banner
    GREEN = 1;              // Green banner level
    YELLOW = 2;             // Yellow banner level  
    RED = 3;                // Red banner level
}

// Banner scope enumeration
enum BannerScope {
    SCOPE_GLOBAL = 0;       // Global banner (screen-wide)
    SCOPE_PARTICIPANT = 1;  // Per-participant banner
}

// Screen banner message for overlay display
message ScreenBanner {
    BannerLevel level = 1;      // Banner level/color
    uint32 ttl_ms = 2;          // Time to live in milliseconds
    string placement = 3;       // Placement hint ("TopCenter" default for global)
    string action_id = 4;       // Unique action identifier
    string scope = 5;           // Action scope ("global" or "participant")
    BannerScope scope_enum = 6; // Banner scope enumeration
    string participant_id = 7;  // Participant ID (for per-participant banners)
    string banner_type = 8;     // Banner type for image selection
    uint64 expiry_timestamp_ms = 9; // Absolute expiry timestamp
}

// Uplink message from Service 5 to backend (client -> server)
message Uplink {
    repeated ParticipantFrame participants = 1;     // Video participant batches (2s cadence)
    optional AudioBatch audio = 2;                  // Audio batch (4s cadence)
    uint64 timestamp_ms = 3;                        // Message timestamp (epoch ms)
    string client_id = 4;                           // Client identifier
    uint64 sequence_number = 5;                     // Message sequence number
}

// Downlink message from backend to Service 5 (server -> client)  
message Downlink {
    optional ScreenBanner screen_banner = 1;        // Screen banner action
    uint64 timestamp_ms = 2;                        // Response timestamp (epoch ms)
    string server_id = 3;                           // Server identifier
    uint64 sequence_number = 4;                     // Response sequence number
    bool received = 5;                              // Confirmation of receipt
    string error_message = 6;                       // Error message if any
}

// Streaming service definition
service StreamingService {
    // Bidirectional streaming for real-time communication
    // Service 5 sends Uplink messages (video every 2s, audio every 4s)
    // Backend responds with Downlink messages (banner actions, confirmations)
    rpc StreamData(stream Uplink) returns (stream Downlink);
    
    // Health check endpoint
    rpc Ping(PingRequest) returns (PingResponse);
}

// Ping request message
message PingRequest {
    uint64 timestamp_ms = 1;        // Request timestamp (epoch ms)
    string client_id = 2;           // Client identifier
}

// Ping response message  
message PingResponse {
    string status = 1;              // Status ("ok" or error)
    uint64 server_time_ms = 2;      // Server timestamp (epoch ms)
    string version = 3;             // Server version
    uint64 request_timestamp_ms = 4; // Echo of request timestamp
}